// riverbed-appinternals/lib/trace-cls.js
/* ****************************
** Copyright (c) 2015        **
** Riverbed Technology, Inc. **
** All Rights Reserved.      **
** ************************* */

"use strict";

// *****************************
// CLS/Namespace Module
// *****************************

// required modules
var cls = require('continuation-local-storage'),
	traceOutputApi = require('./trace-output-api');

// namespace for tracing data
var riverbedTraceNamespace = null;

var getTracingNamespace = exports.getTracingNamespace = function() {

	if(riverbedTraceNamespace == null) {
		riverbedTraceNamespace = cls.createNamespace('riverbed trace');
	}

	return riverbedTraceNamespace;
};

var set = exports.set = function(name,val){
	return getTracingNamespace().set(name,val);
};

var get = exports.get = function(name){
	return getTracingNamespace().get(name);
};

var bindEmitter = exports.bindEmitter = function(emitter){
	return getTracingNamespace().bindEmitter(emitter);
};

var bindEmitterWithHooks = exports.bindEmitterWithHooks = function(emitter,pre,post){
	return getTracingNamespace().bindEmitterWithHooks(emitter,pre,post);
};

var run = exports.run = function(func) {
	return getTracingNamespace().run(func);
};

var bind = exports.bind = function(func) {
	return getTracingNamespace().bind(func);
};

var getData = exports.getData = function(){
	return get('riverbedTracingData');
};

var setData = exports.setData = function(val) {
	return set('riverbedTracingData',val);
};

// if there's tracing data visible, execute the function on that data
var ifDataDo = exports.ifDataDo = function(loc,func) {
	var tracingData = getData();
	if(tracingData) {
		return func(tracingData);
	} else {
		traceOutputApi.debug(loc,"did not find trace");
	}
};
